app/mixins/move_actions.js

```diff
@@ -0,0 +1,137 @@
+import Ember from "ember";
+import config from "stock/config/environment";
+import AsyncMixin from "./async";
+import _ from "lodash";
+
+const ALLOW_PARTIAL_QTY = config.APP.ALLOW_PARTIAL_OPERATIONS;
+
+/**
+ * Adds the following properties:
+ *
+ * - readyToMove (computed, true if other proerties are valid)
+ * - moveQty
+ * - moveTarget (the package to move)
+ * - moveFrom
+ * - moveTo
+ * - editableQty
+ *
+ * Adds the following actions:
+ *
+ * - beginMove(package, [from], [to])
+ * - completeMove()
+ * - cancelMove()
+ */
+export default Ember.Mixin.create(AsyncMixin, {
+  locationService: Ember.inject.service(),
+
+  editableQty: ALLOW_PARTIAL_QTY,
+
+  moveQty: Ember.computed(
+    "_moveQty",
+    "moveFrom",
+    "moveTarget",
+    "moveTarget.packagesLocations.@each.quantity",
+    {
+      get(k) {
+        return this.get("_moveQty");
+      },
+      set(k, value) {
+        const total = this.quantityAtSource();
+        const qty = Number(value);
+
+        if (!ALLOW_PARTIAL_QTY && qty > 0 && qty !== total) {
+          throw new Error("Partial quantity is not permitted");
+        }
+
+        this.set("_moveQty", qty);
+        return qty;
+      }
+    }
+  ),
+
+  async resolveLocation(loc, opts = {}) {
+    const i18n = this.get("i18n");
+
+    if (loc) {
+      return loc;
+    }
+    const { queryText = "", presetLocations } = opts;
+    const text = i18n.exists(queryText) ? i18n.t(queryText) : queryText;
+
+    return this.get("locationService").userPickLocation({
+      headerText: text,
+      presetLocations: presetLocations
+    });
+  },
+
+  quantityAtSource() {
+    const pkg = this.get("moveTarget");
+    const source = this.get("moveFrom");
```

Comment:
How about naming it as `sourceLocation`?

Reply:
mostly for consistency with the backend (more aesthetically pleasing too imo)